cpp/nvtx3.hpp

```diff
@@ -15,21 +15,67 @@
  */
 #pragma once
 
-#if defined(NVTX3_MINOR_VERSION) and NVTX3_MINOR_VERSION < 0
-#error \
-  "Trying to #include NVTX version 3 in a source file where an older NVTX version has already been included.  If you are not directly using NVTX (the NVIDIA Tools Extension library), you are getting this error because libraries you are using have included different versions of NVTX.  Suggested solutions are: (1) reorder #includes so the newest NVTX version is included first, (2) avoid using the conflicting libraries in the same .c/.cpp file, or (3) update the library using the older NVTX version to use the newer version instead."
-#endif
-
+/* Temporary helper #defines, #undef'ed at end of header */
+#define NVTX3_CPP_VERSION_MAJOR 1
+#define NVTX3_CPP_VERSION_MINOR 0
+
+/* This section handles the decision of whether to provide unversioned symbols.
+ * If NVTX3_CPP_REQUIRE_EXPLICIT_VERSION is #defined, unversioned symbols are
+ * not provided, and explicit-version symbols such as nvtx3::v1::thread_range
+ * and NVTX3_V1_FUNC_RANGE must be used.  By default, the first #include of this
+ * header will define the unversioned symbols such as nvtx3::thread_range and
+ * NVTX3_FUNC_RANGE.  Subsequently including a different major version of this
+ * header without #defining NVTX3_CPP_REQUIRE_EXPLICIT_VERSION triggers an error
+ * since the symbols would conflict.  Subsequently including of a different
+ * minor version within the same major version is allowed. Functionality of
+ * minor versions is cumulative, regardless of include order.
+ * 
+ * Since NVTX3_CPP_REQUIRE_EXPLICIT_VERSION allows all combinations of versions
+ * to coexist without problems within a translation unit, the recommended best
+ * practice for instrumenting header-based libraries with NVTX C++ Wrappers is
+ * is to #define NVTX3_CPP_REQUIRE_EXPLICIT_VERSION before including nvtx3.hpp,
+ * #undef it afterward, and only use explicit-version symbols.  This is not
+ * necessary in common cases, such as instrumenting a standalone application, or
+ * static/shared libraries in the .cpp files or headers private to those projects.
+ */
+#if !defined(NVTX3_CPP_REQUIRE_EXPLICIT_VERSION)
+#define NVTX3_INLINE_THIS_VERSION
+#if !defined(NVTX3_CPP_INLINED_VERSION_MAJOR)
+// First occurrence of this header in the translation unit
+/**
+ * @brief Semantic major version number for NVTX C++ wrappers of unversioned symbols
+ *
+ * Breaking changes may occur between major versions, and different major versions
+ * cannot provide unversioned symbols in the same translation unit (.cpp file).
+ * 
+ * Note: If NVTX3_CPP_REQUIRE_EXPLICIT_VERSION is defined, this macro is not defined.
+ *
+ * Not to be confused with the version number of the NVTX core library.
+ */
+#define NVTX3_CPP_INLINED_VERSION_MAJOR 1  // NVTX3_CPP_VERSION_MAJOR
```

Comment:
Separating the `#define` from the `if` with a comment block made it hard to see what was inside the `if`. 
```suggestion
/**
 * @brief Semantic major/minor version number for NVTX C++ wrappers of unversioned symbols
 *
 * Breaking changes may occur between major versions, and different major versions
 * cannot provide unversioned symbols in the same translation unit (.cpp file).
 * 
 * No breaking changes occur between minor versions -- minor version changes within
 * a major version are purely additive.
 * 
 * Not to be confused with the version number of the NVTX core library.
 *
 * @note: If NVTX3_CPP_REQUIRE_EXPLICIT_VERSION is defined, this macro is not defined.
 *
 */
#if !defined(NVTX3_CPP_INLINED_VERSION_MAJOR) // First occurrence of this header in the translation unit
#define NVTX3_CPP_INLINED_VERSION_MAJOR 1  // NVTX3_CPP_VERSION_MAJOR
#define NVTX3_CPP_INLINED_VERSION_MINOR 0 // NVTX3_CPP_VERSION_MINOR
```

Reply:
I didn't really like having the doxygen bloating the inside of this macro logic either, but is there a way to do what you're suggesting (moving the doxygen comments for NVTX3_CPP_INLINED_VERSION_MAJOR  & _MINOR outside the #if) while still connecting the comments to those macros?  Like, I want to make sure those macros in particular are easy to find the docs for...   in IDEs that show doxygen in tooltips, when searching the doxygen pages for them, etc.  I'm happy to make this change if we can keep that relationship -- I just don't know doxygen enough to make that work.

Reply:
This won't affect Katie and others right?  Maybe we can fix this later?  Maybe you can use a doxygen sub-pages to cover this topic and shorten the doxygen here to mention reading this sub-page for more details?

Reply:
Yeah, in my amend on this comment I am turning off clang format for this chunk, the indenting makes it way easier to read.  While I'd like to move these comments out and refer to these macros with a doxygen reference from somewhere else, I couldn't figure out how to do that yet, so I'm happy leaving it this way for now and fixing that later.  There are no version complications at all from just fixing comments whenever we want.  And the nice thing about the current approach is that we get proper doxygen association for these macros by putting the comments right above them.  (I ran doxygen and checked)

Reply:
OK I fixed the indenting so this is all way more readable.